src/ImageSharp/Formats/Jpeg/Components/Decoder/ColorConverters/JpegColorConverter.FromGrayScaleAvx2.cs

```diff
@@ -0,0 +1,63 @@
+﻿// Copyright (c) Six Labors.
+// Licensed under the Apache License, Version 2.0.
+
+using System;
+using System.Numerics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+#if SUPPORTS_RUNTIME_INTRINSICS
+using System.Runtime.Intrinsics;
+using System.Runtime.Intrinsics.X86;
+using static SixLabors.ImageSharp.SimdUtils;
+#endif
+
+namespace SixLabors.ImageSharp.Formats.Jpeg.Components.Decoder.ColorConverters
+{
+    internal abstract partial class JpegColorConverter
+    {
+        internal sealed class FromGrayscaleAvx2 : Avx2JpegColorConverter
+        {
+            public FromGrayscaleAvx2(int precision)
+                : base(JpegColorSpace.Grayscale, precision)
+            {
+            }
+
+            protected override void ConvertCoreVectorized(in ComponentValues values, Span<Vector4> result)
+            {
+#if SUPPORTS_RUNTIME_INTRINSICS
+                ref Vector256<float> gBase =
+                    ref Unsafe.As<float, Vector256<float>>(ref MemoryMarshal.GetReference(values.Component0));
+
+                ref Vector256<float> resultBase =
+                    ref Unsafe.As<Vector4, Vector256<float>>(ref MemoryMarshal.GetReference(result));
+
+                // Used for the color conversion
+                var scale = Vector256.Create(1 / this.MaximumValue);
+                var one = Vector256.Create(1F);
+
+                // Used for packing
+                ref byte control = ref MemoryMarshal.GetReference(HwIntrinsics.PermuteMaskEvenOdd8x32);
+                Vector256<int> vcontrol = Unsafe.As<byte, Vector256<int>>(ref control);
+
+                int n = result.Length / 8;
+                for (int i = 0; i < n; i++)
+                {
+                    Vector256<float> g = Avx2.PermuteVar8x32(Unsafe.Add(ref gBase, i), vcontrol);
+
+                    g = Avx.Multiply(g, scale);
+
+                    ref Vector256<float> destination = ref Unsafe.Add(ref resultBase, i * 4);
+
+                    destination = Avx.Blend(Avx.Permute(g, 0b00_00_00_00), one, 0b1000_1000);
+                    Unsafe.Add(ref destination, 1) = Avx.Blend(Avx.Permute(g, 0b01_01_01_01), one, 0b1000_1000);
+                    Unsafe.Add(ref destination, 2) = Avx.Blend(Avx.Permute(g, 0b10_10_10_10), one, 0b1000_1000);
+                    Unsafe.Add(ref destination, 3) = Avx.Blend(Avx.Permute(g, 0b11_11_11_11), one, 0b1000_1000);
```

Comment:
If you have any idea what should we try to improve this path, let's add a comment. If you think it's unfixable, consider removing this variant, since it doesn't add noticable perf in comparison to the "Basic" (`Vector4`) variant.

Reply:
@saucecontrol do you have ideas here? Those permutes look expensive.

Reply:
These are in-lane permutes (`vpermilps`), so they're the same cost as a normal shuffle @ 1 cycle latency.  The encoding is 1 byte shorter for `vshufps`, which makes that instruction kind of pointless, but it's not bad.  See: https://stackoverflow.com/a/54170092/4926931

The cross-lane permutes above (`Avx2.PermuteVar8x32`) are higher cost, though.  This might be a case where an SSE implementation is faster, but I haven't tested that one specifically.

Reply:
Thanks!

Reply:
@JimBobSquarePants If you want to get a better idea how the ports + latency and throughput numbers for the various SIMD instructions interact, have a read through chapter 9 of [Agner's optimization guide](https://www.agner.org/optimize/optimizing_assembly.pdf).  Once you have a handle on that, you can model things pretty quickly to get a sense of how they'll perform without necessarily having to write and benchmark them.  The best data for the ports/latency/thoughput currently is on uops.info or in Agner's instruction table PDFs.

I did another crappy diagram™ to show how this one models out with the instruction latency and parallelism:

```
cycle ______input_______
1     vmovups
2     ...
3     ...
4        => vpermd
5           ...
6           ...
7              => vmulps
8                 ...
9                 ...
10                ...       ________out0_______    ________out1_______    ________out2_______    _______out3_______
11                       => vpermilps                     
12                             => vblendps         vpermilps
13                                   => vmovups       => vblendps         vpermilps
14                                      ...                 => vmovups       => vblendps         vpermilps
15                                      ...                    ...                 => vmovups      => vblendps
16                                                             ...                    ...                => vmovups
17                                                                                    ...                    ...
18                                                                                                           ...
```

So ignoring loop overhead, and assuming 32-byte aligned data and perfect cache usage, the SIMD code models out to 18 cycles per iteration for 8 pixels.  The key thing here is that it takes 10 cycles to prepare the data before the output shuffling can start, and then the 4 outputs parallelize nicely, doing the rest of the work in 8 cycles.

If you were to take 2 input vectors at a time, you'd have the same 10 cycles latency before your first shuffle, but the second vector is ready 1 cycle after that, and the rest can be paralleled in the same way.  You end up with 10 cycles for the input prep and 12 for the shuffle/blend/store, making it 22 cycles for 16 pixels.  You could go even further and get 26 cycles for 24 pixels by unrolling the input by 3, or 30 cycles for 32 pixels, etc.  At some point, you run out of registers, your loop body gets too big for the uop cache, or you simply have chunks so big your remainder gets out of hand, so there's a balance to be found somewhere.

If it were me, I'd unroll by at least 2, swap the order of the multiply and permute8x32, and change the `vpermilps` to `vshufps` to minimize the loop body code size.  That should give enough gain to make the AVX2 implementation worthwhile.

Reply:
So I changed the input size of the benchmarks from 128 to 512 and re-ran it with the following changes:

* `SimdVectorAvx2` unchanged
* `SimdVectorAvx2v2` swapped multiply and permute and changed `vpermilps` to `vshufps`
* `SimdVectorAvx2By2` All suggested changes (`SimdVectorAvx2v2` + unrolled by 2)

|            Method |        Job |       Runtime |     Mean |     Error |   StdDev | Ratio | RatioSD |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------ |----------- |-------------- |---------:|----------:|---------:|------:|--------:|-------:|------:|------:|----------:|
|            Scalar | Job-KDKLQO | .NET Core 3.1 | 400.3 ns | 107.51 ns |  5.89 ns |  1.00 |    0.00 | 0.0048 |     - |     - |      32 B |
|    SimdVectorAvx2 | Job-KDKLQO | .NET Core 3.1 | 240.7 ns |  36.36 ns |  1.99 ns |  0.60 |    0.01 | 0.0062 |     - |     - |      40 B |
|  SimdVectorAvx2v2 | Job-KDKLQO | .NET Core 3.1 | 233.2 ns |   4.93 ns |  0.27 ns |  0.58 |    0.01 | 0.0062 |     - |     - |      40 B |
| SimdVectorAvx2By2 | Job-KDKLQO | .NET Core 3.1 | 223.9 ns |  55.07 ns |  3.02 ns |  0.56 |    0.00 | 0.0062 |     - |     - |      40 B |


--> I'll apply your suggested improvements but ignore unrolling for now, as the gains are relatively small vs. code size increase.